invokeai/backend/model_manager/loader.py

```diff
@@ -0,0 +1,301 @@
+
+
+class ModelLoadBase(ABC):
+    """Abstract base class for a model loader which works with the ModelConfigStore backend."""
+
+    @abstractmethod
+    def get_model(self, key: str, submodel_type: Optional[SubModelType] = None) -> ModelInfo:
+        """
+        Return a model given its key.
+
+        Given a model key identified in the model configuration backend,
+        return a ModelInfo object that can be used to retrieve the model.
+
+        :param key: model key, as known to the config backend
+        :param submodel_type: an ModelType enum indicating the portion of
+               the model to retrieve (e.g. ModelType.Vae)
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def store(self) -> ModelConfigStore:
+        """Return the ModelConfigStore object that supports this loader."""
+        pass
+
+    @property
+    @abstractmethod
+    def installer(self) -> ModelInstallBase:
+        """Return the ModelInstallBase object that supports this loader."""
+        pass
+
+    @property
+    @abstractmethod
+    def logger(self) -> Logger:
+        """Return the current logger."""
+        pass
+
+    @property
+    def config(self) -> InvokeAIAppConfig:
+        """Return the config object used by this installer."""
+        pass
+
+    @property
+    @abstractmethod
+    def queue(self) -> DownloadQueueBase:
+        """Return the download queue object used by this object."""
+        pass
+
+    @abstractmethod
+    def collect_cache_stats(self, cache_stats: CacheStats):
+        """Replace cache statistics."""
+        pass
+
+    @abstractmethod
+    def resolve_model_path(self, path: Union[Path, str]) -> Path:
+        """Turn a potentially relative path into an absolute one in the models_dir."""
+        pass
+
+    @property
+    @abstractmethod
+    def precision(self) -> str:
+        """Return 'float32' or 'float16'."""
+        pass
+
+    @abstractmethod
+    def sync_to_config(self):
+        """Reinitialize the store to sync in-memory and in-disk versions."""
+        pass
+
+
+class ModelLoad(ModelLoadBase):
+    """Implementation of ModelLoadBase."""
+
+    _app_config: InvokeAIAppConfig
+    _store: ModelConfigStore
+    _installer: ModelInstallBase
+    _cache: ModelCache
+    _logger: Logger
+    _cache_keys: dict
+    _models_file: Path
+
+    def __init__(self, config: InvokeAIAppConfig, event_handlers: List[DownloadEventHandler] = []):
+        """
+        Initialize ModelLoad object.
+
+        :param config: The app's InvokeAIAppConfig object.
+        """
+        if config.model_conf_path and config.model_conf_path.exists():
+            models_file = config.model_conf_path
+        else:
+            models_file = config.root_path / "configs/models3.yaml"
+        try:
+            store = get_config_store(models_file)
+        except ConfigFileVersionMismatchException:
+            migrate_models_store(config)
+            store = get_config_store(models_file)
```

Comment:
Do we want to default to yaml model storage? Probably oughta have a discussion around this. I'm all for putting it in the database and not letting users use a yaml file at all.

It's much simpler if we only need to maintain a single implementation and we don't want to have users editing config files if we can help it. Yaml in particular is tricky due to semantic whitespace, very easy to mess up.

Reply:
For development and debugging, it's much easier to use the yaml file because I can see what's going on easily. For production, we should default to using the db, and I propose we use the same db that the image and board storage use.

With respect to locking and connections, I thought that concurrent connections were supported by SQLite3 (https://sqlite.org/faq.html#q5)? I'll go ahead and rewrite the __init__ routine to accept the shared connection and lock the way that the image and board services use.

With respect to merge timing, this depends on your schedule for making the required frontend changes to support the altered API. It doesn't look like a big deal one way or the other.

Reply:
SQLite supports concurrent reads, but once there is a write in progress, no reads (or writes) can be made.

There were issues with our multiple threads accessing the db at once that were resolved by using a single connection. Haven't had any issues since implementing that.